src/script/descriptor.cpp

```diff
@@ -829,6 +839,45 @@ class WSHDescriptor final : public DescriptorImpl
     bool IsSingleType() const final { return true; }
 };
 
+/** A parsed tr(...) descriptor. */
+class TRDescriptor final : public DescriptorImpl
+{
+    std::vector<int> m_depths;
+protected:
+    std::vector<CScript> MakeScripts(const std::vector<CPubKey>& keys, Span<const CScript> scripts, FlatSigningProvider& out) const override
+    {
+        TaprootBuilder builder;
+        assert(m_depths.size() == scripts.size());
+        for (size_t pos = 0; pos < m_depths.size(); ++pos) {
+            builder.Add(m_depths[pos], scripts[pos], TAPROOT_LEAF_TAPSCRIPT);
```

Comment:
given that this is just iterating in order, it seems like we should be able to directly feed the TaprootBuilder from the tr parsing code and do away with the intermediate vectors?

Reply:
No, unfortunately this happens at a different layers. TaprootBuilder can only be invoked with concrete keys; those don't exist yet at parsing time.

Reply:
I see :(

I guess I'll mark it as a personal todo (no need to upset a working internal representation) to brainstorm if there's a more obvious way to represent the tree than the ordered depth annotations (even though it's equivalent and unambiguous, I still find it unintuitive)

Reply:
I like linear data structures that don't need recursion to process, when unnecessary.

Reply:
Yeah, I agree with that 100%. What I was thinking was a vec<uint128> where each bit represents a left/right on the tree. That way you can do a de-duplicate check to make sure each node is only occupied once. Then the order of the vec itself is no longer significant; it should not need to be preserved because you can sort them by value (if I'm not wrong) to get the insertion order. Depth could be computed by looking at your nearest neighbor and computing the "divergence" (or by encoding an extra byte to make it explicit -- unclear what is best here). 

With a sorted vec, this also enables some nicesties such as using std:: binary searches for O(log(n)) lookup by tree position.

Reply:
@roconnor-blockstream may also like that this format should be:
1. concatenative, where concat has operations that are O(n)
    - merge sort then either deduplicate check or run through and add an appropriate offset to branches
 2.  "normalizable" (shifting things down to the smallest next number) and "prefixable" (adding a constant to all numbers to preserve structure.

Reply:
Why do the properties of this internal, temporary, data structure matter? It does what it is supposed to do. If we ever need other operations on it, something else can be considered.

Reply:
I find the code really hard to understand. Conceivably if we want people to review/audit it, a conceptually simpler representation & set of algorithms should be easier to understand? The above suggestion may or may not be it, but I haven't been able to fully work out how the algorithms this PR is currently using work even after staring at them for a bit.

the properties matter less.

Reply:
I'll try to add more comments.

Reply:
Added a big comment explaining the data structure in script/standard.h. WDYT?

Reply:
Marking this as resolved. Feel free to comment if the result still isn't clear.